pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -1107,106 +1292,95 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListenerForPortOrUDS(l
+	//  Type 3 can be resolved by appending service to existing services
+	//  Type 4 can be resolved by merging HTTP filter chain with TCP filter chain
+	//  Type 5 can be resolved by merging TCP filter chains
+	//  Type 6 can be resolved by merging TCP filter chains
+	//  Type 7 can be resolved by appending service to existing services
+	//  Type 8 can be resolved by merging TCP filter chains
+	//  Type 9 can be resolved by merging TCP and HTTP filter chains
+
+	switch conflictType {
+	case NoConflict:
+		if currentListenerEntry != nil {
+			currentListenerEntry.listener.FilterChains = mergeTCPFilterChains(mutable.Listener.FilterChains,
+				pluginParams, listenerMapKey, listenerMap, node)
+		} else {
+			listenerMap[listenerMapKey] = &outboundListenerEntry{
+				services:    []*model.Service{pluginParams.Service},
+				servicePort: pluginParams.Port,
+				bind:        listenerOpts.bind,
+				listener:    mutable.Listener,
+				protocol:    pluginParams.Port.Protocol,
 			}
 		}
-		currentListenerEntry.listener.FilterChains = newFilterChains
-	} else {
+
+	case HTTPOverHTTP, HTTPOverAuto:
+		// Append service only
+		currentListenerEntry.services = append(currentListenerEntry.services, pluginParams.Service)
+	case HTTPOverTCP:
+		// Merge HTTP filter chain to TCP filter chain
+		currentListenerEntry.listener.FilterChains = mergeFilterChains(mutable.Listener.FilterChains, currentListenerEntry.listener.FilterChains)
+		currentListenerEntry.protocol = protocol.Unsupported
+		currentListenerEntry.listener.ListenerFilters =
+			append(currentListenerEntry.listener.ListenerFilters, &listener.ListenerFilter{Name: envoyListenerHTTPInspector})
+		currentListenerEntry.services = append(currentListenerEntry.services, pluginParams.Service)
+
+	case TCPOverHTTP:
+		// Merge TCP filter chain to HTTP filter chain
+		currentListenerEntry.listener.FilterChains = mergeFilterChains(currentListenerEntry.listener.FilterChains, mutable.Listener.FilterChains)
+		currentListenerEntry.protocol = protocol.Unsupported
+		currentListenerEntry.listener.ListenerFilters =
+			append(currentListenerEntry.listener.ListenerFilters, &listener.ListenerFilter{Name: envoyListenerHTTPInspector})
+	case TCPOverTCP:
+		// Merge two TCP filter chains. HTTP filter chain will not conflict with TCP filter chain because HTTP filter chain match for
+		// HTTP filter chain is different from TCP filter chain's.
+		currentListenerEntry.listener.FilterChains = mergeTCPFilterChains(mutable.Listener.FilterChains,
+			pluginParams, listenerMapKey, listenerMap, node)
+	case TCPOverAuto:
+		// Merge two TCP filter chains. HTTP filter chain will not conflict with TCP filter chain because HTTP filter chain match for
+		// HTTP filter chain is different from TCP filter chain's.
+		currentListenerEntry.listener.FilterChains = mergeTCPFilterChains(mutable.Listener.FilterChains,
+			pluginParams, listenerMapKey, listenerMap, node)
+
+	case AutoOverHTTP:
```

Comment:
this is effectively same as tcp over http as auto has both tcp+http. so you have to add the filter chain for tcp to this listener.

Reply:
Yes. The code is a little bit tricky. Instead of merging TCP+HTTP to HTTP, I merged HTTP to the TCP+HTTP filter chains. I only need to append services and replace the entry in listener map.